# Analysis of European Soccer, 2008-2016.

<img src="./images/soccer.PNG" alt="European Soccer Players">

## Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#importing">Importing The Datasets</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
</ul>

<a id='intro'></a>
## Introduction 

This dataset, taken from a Kaggle dataset repository, consists of 11 european countries and their respective championships or leagues. There are 25,000 matches plus in the dataset and more than 10,000 players. But for our analysis, we will concentrate on just the three four leagues in Europe: The Premiership in England, the Bundesliga in Germany, La liga in Spain, and Series A in Italy.

<a id='importing'></a>
## Importing the dataset

The dataset is an SQLite database with 7 tables - Country, League, Match, Player, Player_Attributes, Team, and Team_Attributes. To use the data, we need to convert the tables to a pandas dataframe so we could clean it and generate insights. 

The code to convert all the tables to csv was found from a stackoverflow post. The link to the post is in the references section of this notebook. 

In [1]:
import pandas as pd
import sqlite3


In [2]:
# # uncomment this section if this is your first time here. These codes create the csv files 

# # establish sqlite connection
# conn = sqlite3.connect('database.sqlite')
# c = conn.cursor()

# # loop through the database tables and extract them. 
# for table in c.execute("SELECT name FROM sqlite_master WHERE type='table';").fetchall():
#     t = table[0]
#     df = pd.read_sql('SELECT * from ' + t, conn)
#     df.to_csv('./data/' + t + '.csv', index=False)

The CSV files generated are large files, so I gitignored them while pushing to Github. You have to run the above python code yourself to be able to run this notebook locally. 

In [3]:
# now let's import each of the datasets to a dataframe
# the country dataset
country_df = pd.read_csv("./data_csvs/Country.csv")
country_df.head(5)

id     name
0      1  Belgium
1   1729  England
2   4769   France
3   7809  Germany
4  10257    Italy

In [4]:
# the league dataset
league_df = pd.read_csv("./data_csvs/League.csv")
league_df.head()

id  country_id                    name
0      1           1  Belgium Jupiler League
1   1729        1729  England Premier League
2   4769        4769          France Ligue 1
3   7809        7809   Germany 1. Bundesliga
4  10257       10257           Italy Serie A

In [5]:
# the match dataset
match_df = pd.read_csv("./data_csvs/Match.csv")
# transpose the head because of so many columns
match_df.head().T

0          1          2          3          4
id                  1          2          3          4          5
country_id          1          1          1          1          1
league_id           1          1          1          1          1
season      2008/2009  2008/2009  2008/2009  2008/2009  2008/2009
stage               1          1          1          1          1
...               ...        ...        ...        ...        ...
GBD              3.25       3.25        3.2       3.75        3.5
GBA               4.0       3.75        2.5        5.5       1.65
BSH              1.73       1.91        2.3       1.44       4.75
BSD               3.4       3.25        3.2       3.75        3.3
BSA               4.2        3.6       2.75        6.5       1.67

[115 rows x 5 columns]

Let's get a sense of the columns in the match dataframe by outputing them

In [6]:
match_list = match_df.columns
print(match_list[:30]) # first thirty
print(match_list[30:60]) # next thirty
print(match_list[60:95]) # next thirty five
print(match_list[95:]) # last 20

Index(['id', 'country_id', 'league_id', 'season', 'stage', 'date',
       'match_api_id', 'home_team_api_id', 'away_team_api_id',
       'home_team_goal', 'away_team_goal', 'home_player_X1', 'home_player_X2',
       'home_player_X3', 'home_player_X4', 'home_player_X5', 'home_player_X6',
       'home_player_X7', 'home_player_X8', 'home_player_X9', 'home_player_X10',
       'home_player_X11', 'away_player_X1', 'away_player_X2', 'away_player_X3',
       'away_player_X4', 'away_player_X5', 'away_player_X6', 'away_player_X7',
       'away_player_X8'],
      dtype='object')
Index(['away_player_X9', 'away_player_X10', 'away_player_X11',
       'home_player_Y1', 'home_player_Y2', 'home_player_Y3', 'home_player_Y4',
       'home_player_Y5', 'home_player_Y6', 'home_player_Y7', 'home_player_Y8',
       'home_player_Y9', 'home_player_Y10', 'home_player_Y11',
       'away_player_Y1', 'away_player_Y2', 'away_player_Y3', 'away_player_Y4',
       'away_player_Y5', 'away_player_Y6', 'away_player_Y7', '

We'll further analyze the match dataframe soon. Let's import the rest of the datasets

In [7]:
player_attributes_df = pd.read_csv("./data_csvs/Player_Attributes.csv")
player_attributes_df.head().T

0                    1  \
id                                     1                    2   
player_fifa_api_id                218353               218353   
player_api_id                     505942               505942   
date                 2016-02-18 00:00:00  2015-11-19 00:00:00   
overall_rating                      67.0                 67.0   
potential                           71.0                 71.0   
preferred_foot                     right                right   
attacking_work_rate               medium               medium   
defensive_work_rate               medium               medium   
crossing                            49.0                 49.0   
finishing                           44.0                 44.0   
heading_accuracy                    71.0                 71.0   
short_passing                       61.0                 61.0   
volleys                             44.0                 44.0   
dribbling                           51.0                 51.0   
curve                               45.0                 45.0   
free_kick_accuracy                  39.0                 39.0   
long_passing                        64.0                 64.0   
ball_control                        49.0                 49.0   
acceleration                        60.0                 60.0   
sprint_speed                        64.0                 64.0   
agility                             59.0                 59.0   
reactions                           47.0                 47.0   
balance                             65.0                 65.0   
shot_power                          55.0                 55.0   
jumping                             58.0                 58.0   
stamina                             54.0                 54.0   
strength                            76.0                 76.0   
long_shots                          35.0                 35.0   
aggression                          71.0                 71.0   
interceptions                       70.0                 70.0   
positioning                         45.0                 45.0   
vision                              54.0                 54.0   
penalties                           48.0                 48.0   
marking                             65.0                 65.0   
standing_tackle                     69.0                 69.0   
sliding_tackle                      69.0                 69.0   
gk_diving                            6.0                  6.0   
gk_handling                         11.0                 11.0   
gk_kicking                          10.0                 10.0   
gk_positioning                       8.0                  8.0   
gk_reflexes                          8.0                  8.0   

                                       2                    3  \
id                                     3                    4   
player_fifa_api_id                218353               218353   
player_api_id                     505942               505942   
date                 2015-09-21 00:00:00  2015-03-20 00:00:00   
overall_rating                      62.0                 61.0   
potential                           66.0                 65.0   
preferred_foot                     right                right   
attacking_work_rate               medium               medium   
defensive_work_rate               medium               medium   
crossing                            49.0                 48.0   
finishing                           44.0                 43.0   
heading_accuracy                    71.0                 70.0   
short_passing                       61.0                 60.0   
volleys                             44.0                 43.0   
dribbling                           51.0                 50.0   
curve                               45.0                 44.0   
free_kick_accuracy                  39.0                 38.0   
long_passing                        64.0                 63.0   
ball_con

In [8]:
# the player dataframe
player_df = pd.read_csv("./data_csvs/Player.csv")
player_df.head()

id  player_api_id         player_name  player_fifa_api_id  \
0   1         505942  Aaron Appindangoye              218353   
1   2         155782     Aaron Cresswell              189615   
2   3         162549         Aaron Doran              186170   
3   4          30572       Aaron Galindo              140161   
4   5          23780        Aaron Hughes               17725   

              birthday  height  weight  
0  1992-02-29 00:00:00  182.88     187  
1  1989-12-15 00:00:00  170.18     146  
2  1991-05-13 00:00:00  170.18     163  
3  1982-05-08 00:00:00  182.88     198  
4  1979-11-08 00:00:00  182.88     154

In [9]:
# the team attributes data set
team_attributes_df = pd.read_csv("./data_csvs/Team_Attributes.csv")
team_attributes_df.head().T

0                    1  \
id                                                1                    2   
team_fifa_api_id                                434                  434   
team_api_id                                    9930                 9930   
date                            2010-02-22 00:00:00  2014-09-19 00:00:00   
buildUpPlaySpeed                                 60                   52   
buildUpPlaySpeedClass                      Balanced             Balanced   
buildUpPlayDribbling                            NaN                 48.0   
buildUpPlayDribblingClass                    Little               Normal   
buildUpPlayPassing                               50                   56   
buildUpPlayPassingClass                       Mixed                Mixed   
buildUpPlayPositioningClass               Organised            Organised   
chanceCreationPassing                            60                   54   
chanceCreationPassingClass                   Normal               Normal   
chanceCreationCrossing                           65                   63   
chanceCreationCrossingClass                  Normal               Normal   
chanceCreationShooting                           55                   64   
chanceCreationShootingClass                  Normal               Normal   
chanceCreationPositioningClass            Organised            Organised   
defencePressure                                  50                   47   
defencePressureClass                         Medium               Medium   
defenceAggression                                55                   44   
defenceAggressionClass                        Press                Press   
defenceTeamWidth                                 45                   54   
defenceTeamWidthClass                        Normal               Normal   
defenceDefenderLineClass                      Cover                Cover   

                                                  2                    3  \
id                                                3                    4   
team_fifa_api_id                                434                   77   
team_api_id                                    9930                 8485   
date                            2015-09-10 00:00:00  2010-02-22 00:00:00   
buildUpPlaySpeed                                 47                   70   
buildUpPlaySpeedClass                      Balanced                 Fast   
buildUpPlayDribbling                           41.0                  NaN   
buildUpPlayDribblingClass                    Normal               Little   
buildUpPlayPassing                               54                   70   
buildUpPlayPassingClass                       Mixed                 Long   
buildUpPlayPositioningClass               Organised            Organised   
chanceCreationPassing                            54                   70   
chanceCreationPassingClass                   Normal                Risky   
chanceCreationCrossing                           63                   70   
chanceCreationCrossingClass                  Normal                 Lots   
chanceCreationShooting                           64                   70   
chanceCreationShootingClass                  Normal                 Lots   
chanceCreationPositioningClass            Organised            Organised   
defencePressure                                  47                   60   
defencePressureClass                         Medium               Medium   
defenceAggression                                44                   70   
defenceAggressionClass                        Press               Double   
defenceTeamWidth                                 54                   70   
defenceTeamWidthClass                        Normal                 Wide   
defenceDefenderLineClass                      Cover                Cover   

                                                  4  
id                                        

In [10]:
# finally, the team data frame
team_df = pd.read_csv("./data_csvs/Team.csv")
team_df.head()

id  team_api_id  team_fifa_api_id     team_long_name team_short_name
0   1         9987             673.0           KRC Genk             GEN
1   2         9993             675.0       Beerschot AC             BAC
2   3        10000           15005.0   SV Zulte-Waregem             ZUL
3   4         9994            2007.0   Sporting Lokeren             LOK
4   5         9984            1750.0  KSV Cercle Brugge             CEB

To be honest, what I can analyze from the dataset is limited because I lack access to a subject matter expert for the columns in some of the dataframes. But would do my best to extract insights from the much I know about the game of soccer and how the game is organized. 

<a id='wrangling'></a>
## Data Wrangling

This section involves assessing and cleaning the data. We have 7 data frames and they need to be wrangled. 

I will start with the country data frame. 


In [11]:
# getting the dimensions of the country_df data frame
f'There are {country_df.shape[0]} samples and {country_df.shape[1]} columns'

'There are 11 samples and 2 columns'

So we have 11 countries in the data frame. We need only four: England, Germany, Spain, and Italy. 

In [12]:
# checking the data types for inconsistencies
country_df.dtypes

id       int64
name    object
dtype: object

There are no inconsistencies in the data types. 

In [13]:
# check for nulls or missing values
country_df.isnull().sum()

id      0
name    0
dtype: int64

No nulls or missing values in the country data frame. 

In [14]:
# check for duplicates
sum(country_df.duplicated())

0

There are no duplicates. 

In [15]:
# check for number of unique values in each column. 
country_df.nunique()

id      11
name    11
dtype: int64

This confirms what we already know that the country data frame contains 11 countries. 

Next is the League dataframe, league_df.

In [16]:
# check for the dimension of the league dataframe
f'There are {league_df.shape[0]} samples and {league_df.shape[1]} columns'

'There are 11 samples and 3 columns'

The 11 samples refer to the 11 countries in the data frame. We are interested in 4. 

In [17]:
# checking for data types consistency
league_df.dtypes

id             int64
country_id     int64
name          object
dtype: object

The data types are consistent with what is expected. 

In [18]:
# check for nulls
league_df.isnull().sum()

id            0
country_id    0
name          0
dtype: int64

There are no missing data in the leagues data frame. 

In [19]:
# check for duplicates
sum(league_df.duplicated())

0

There are no duplicate rows. 

In [20]:
# check for the unique values in each column
league_df.nunique()

id            11
country_id    11
name          11
dtype: int64

This confirms that in the data frame there are 11 countries and 11 leagues.

Next is the match data frame. 

In [21]:
# check for dimensions
f'There are {match_df.shape[0]} samples and {match_df.shape[1]} columns'

'There are 25979 samples and 115 columns'

For the 9 years of collected data, we can see that there are 25,979 matches data collected with 115 columns. 

In [22]:
# checking for datatype consistency
match_df.dtypes

id              int64
country_id      int64
league_id       int64
season         object
stage           int64
               ...   
GBD           float64
GBA           float64
BSH           float64
BSD           float64
BSA           float64
Length: 115, dtype: object

The data types are consistent, so we don't need to work on them. 

In [23]:
# checking for missing data and nulls
match_nulls = match_df.isnull().sum()
for index, value in match_nulls.items():
    if value > 0:
        print(index, value)

home_player_X1 1821
home_player_X2 1821
home_player_X3 1832
home_player_X4 1832
home_player_X5 1832
home_player_X6 1832
home_player_X7 1832
home_player_X8 1832
home_player_X9 1832
home_player_X10 1832
home_player_X11 1832
away_player_X1 1832
away_player_X2 1832
away_player_X3 1832
away_player_X4 1832
away_player_X5 1832
away_player_X6 1832
away_player_X7 1832
away_player_X8 1832
away_player_X9 1833
away_player_X10 1833
away_player_X11 1839
home_player_Y1 1821
home_player_Y2 1821
home_player_Y3 1832
home_player_Y4 1832
home_player_Y5 1832
home_player_Y6 1832
home_player_Y7 1832
home_player_Y8 1832
home_player_Y9 1832
home_player_Y10 1832
home_player_Y11 1832
away_player_Y1 1832
away_player_Y2 1832
away_player_Y3 1832
away_player_Y4 1832
away_player_Y5 1832
away_player_Y6 1832
away_player_Y7 1832
away_player_Y8 1832
away_player_Y9 1833
away_player_Y10 1833
away_player_Y11 1839
home_player_1 1224
home_player_2 1315
home_player_3 1281
home_player_4 1323
home_player_5 1316
home_player_6 132

The following columns of interest have missing values: goal, shoton, shotoff, foulcommit, card, cross, corner, and possession. These are values that cannot be imputed for a match so they will be left as-is. 

In [24]:
# check for duplicates
sum(match_df.duplicated())

0

There are no duplicated matches in the match data frame. 

Each match is unique so I expect a huge numer of unique values for columns. So it will not be investigated. 

In [25]:
# getting useful descriptive statistics in the match data frame
match_df.describe()

id    country_id     league_id         stage  match_api_id  \
count  25979.000000  25979.000000  25979.000000  25979.000000  2.597900e+04   
mean   12990.000000  11738.630317  11738.630317     18.242773  1.195429e+06   
std     7499.635658   7553.936759   7553.936759     10.407354  4.946279e+05   
min        1.000000      1.000000      1.000000      1.000000  4.831290e+05   
25%     6495.500000   4769.000000   4769.000000      9.000000  7.684365e+05   
50%    12990.000000  10257.000000  10257.000000     18.000000  1.147511e+06   
75%    19484.500000  17642.000000  17642.000000     27.000000  1.709852e+06   
max    25979.000000  24558.000000  24558.000000     38.000000  2.216672e+06   

       home_team_api_id  away_team_api_id  home_team_goal  away_team_goal  \
count      25979.000000      25979.000000    25979.000000    25979.000000   
mean        9984.371993       9984.475115        1.544594        1.160938   
std        14087.453758      14087.445135        1.297158        1.142110   
min         1601.000000       1601.000000        0.000000        0.000000   
25%         8475.000000       8475.000000        1.000000        0.000000   
50%         8697.000000       8697.000000        1.000000        1.000000   
75%         9925.000000       9925.000000        2.000000        2.000000   
max       274581.000000     274581.000000       10.000000        9.000000   

       home_player_X1  ...           SJA           VCH           VCD  \
count    24158.000000  ...  17097.000000  22568.000000  22568.000000   
mean         0.999586  ...      4.622343      2.668107      3.899048   
std          0.022284  ...      3.632164      1.928753      1.248221   
min          0.000000  ...      1.100000      1.030000      1.620000   
25%          1.000000  ...      2.500000      1.700000      3.300000   
50%          1.000000  ...      3.500000      2.150000      3.500000   
75%          1.000000  ...      5.250000      2.800000      4.000000   
max          2.000000  ...     41.000000     36.000000     26.000000   

                VCA           GBH           GBD           GBA           BSH  \
count  22568.000000  14162.000000  14162.000000  14162.000000  14161.000000   
mean       4.840281      2.498764      3.648189      4.353097      2.497894   
std        4.318338      1.489299      0.867440      3.010189      1.507793   
min        1.080000      1.050000      1.450000      1.120000      1.040000   
25%        2.550000      1.670000      3.200000      2.500000      1.670000   
50%        3.500000      2.100000      3.300000      3.400000      2.100000   
75%        5.400000      2.650000      3.750000      5.000000      2.620000   
max       67.000000     21.000000     11.000000     34.000000     17.000000   

                BSD           BSA  
count  14161.000000  14161.000000  
mean       3.660742      4.405663  
std        0.868272      3.189814  
min        1.330000      1.120000  
25%        3.250000      2.500000  
50%        3.400000      3.400000  
75%        3.750000      5.000000  
max       13.000000     34.000000  

[8 rows x 105 columns]

Just to take a few. The highest goal scored by a home team is 10 in a match and the highest by the away team is 9 in a match. The mean goals for both home and away teams are 1. 

Now, we'll consider the player_attributes dataframe. 

In [26]:
# checking the dimensions
f'There are {player_attributes_df.shape[0]} samples and {player_attributes_df.shape[1]} columns.'

'There are 183978 samples and 42 columns.'

In [27]:
# checking for data consistency
player_attributes_df.dtypes

id                       int64
player_fifa_api_id       int64
player_api_id            int64
date                    object
overall_rating         float64
potential              float64
preferred_foot          object
attacking_work_rate     object
defensive_work_rate     object
crossing               float64
finishing              float64
heading_accuracy       float64
short_passing          float64
volleys                float64
dribbling              float64
curve                  float64
free_kick_accuracy     float64
long_passing           float64
ball_control           float64
acceleration           float64
sprint_speed           float64
agility                float64
reactions              float64
balance                float64
shot_power             float64
jumping                float64
stamina                float64
strength               float64
long_shots             float64
aggression             float64
interceptions          float64
positioning            float64
vision  

In [28]:
player_attributes_df['defensive_work_rate'].unique()

array(['medium', 'high', 'low', '_0', nan, '5', 'ean', 'o', '1', 'ormal',
       '7', '2', '8', '4', 'tocky', '0', '3', '6', '9', 'es'],
      dtype=object)

There are some inconsistent datatypes in the player_attributes dataframe. 
1. Date is characterized as a string. It has to be converted to a datatime datatype. 
2. Defensive_work_rate has values that are both strings and numbers. It is characterized as a string. Since there is no data dictionary to help do a mapping, I will leave it as it is.  

In [29]:
# check for nulls and missing data 
player_attributes_df.isnull().sum()

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
overall_rating          836
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle     

Some players don't have attribute values assigned to them. Because attributes are player specific and can change over time, i.e dynamic, I will leave the missing attributes as they are. 

In [30]:
# check for duplicates
sum(player_attributes_df.duplicated())

0

There are no duplicates in the data frame. That's good. 

In [31]:
# check for number of unique values in each column
player_attributes_df.nunique()

id                     183978
player_fifa_api_id      11062
player_api_id           11060
date                      197
overall_rating             61
potential                  56
preferred_foot              2
attacking_work_rate         8
defensive_work_rate        19
crossing                   95
finishing                  97
heading_accuracy           96
short_passing              95
volleys                    93
dribbling                  97
curve                      92
free_kick_accuracy         97
long_passing               95
ball_control               93
acceleration               86
sprint_speed               85
agility                    81
reactions                  78
balance                    81
shot_power                 96
jumping                    79
stamina                    84
strength                   82
long_shots                 96
aggression                 91
interceptions              96
positioning                95
vision                     97
penalties 

From the unique values in each column, we can see that the data frame contains information on 11,062 players in Europe for the period under consideration. Some columns also have a great variety of unique values. 

In [32]:
# let's take descriptive statistics
player_attributes_df.describe()

id  player_fifa_api_id  player_api_id  overall_rating  \
count  183978.00000       183978.000000  183978.000000   183142.000000   
mean    91989.50000       165671.524291  135900.617324       68.600015   
std     53110.01825        53851.094769  136927.840510        7.041139   
min         1.00000            2.000000    2625.000000       33.000000   
25%     45995.25000       155798.000000   34763.000000       64.000000   
50%     91989.50000       183488.000000   77741.000000       69.000000   
75%    137983.75000       199848.000000  191080.000000       73.000000   
max    183978.00000       234141.000000  750584.000000       94.000000   

           potential       crossing      finishing  heading_accuracy  \
count  183142.000000  183142.000000  183142.000000     183142.000000   
mean       73.460353      55.086883      49.921078         57.266023   
std         6.592271      17.242135      19.038705         16.488905   
min        39.000000       1.000000       1.000000          1.000000   
25%        69.000000      45.000000      34.000000         49.000000   
50%        74.000000      59.000000      53.000000         60.000000   
75%        78.000000      68.000000      65.000000         68.000000   
max        97.000000      95.000000      97.000000         98.000000   

       short_passing        volleys  ...         vision      penalties  \
count  183142.000000  181265.000000  ...  181265.000000  183142.000000   
mean       62.429672      49.468436  ...      57.873550      55.003986   
std        14.194068      18.256618  ...      15.144086      15.546519   
min         3.000000       1.000000  ...       1.000000       2.000000   
25%        57.000000      35.000000  ...      49.000000      45.000000   
50%        65.000000      52.000000  ...      60.000000      57.000000   
75%        72.000000      64.000000  ...      69.000000      67.000000   
max        97.000000      93.000000  ...      97.000000      96.000000   

             marking  standing_tackle  sliding_tackle      gk_diving  \
count  183142.000000    183142.000000   181265.000000  183142.000000   
mean       46.772242        50.351257       48.001462      14.704393   
std        21.227667        21.483706       21.598778      16.865467   
min         1.000000         1.000000        2.000000       1.000000   
25%        25.000000        29.000000       25.000000       7.000000   
50%        50.000000        56.000000       53.000000      10.000000   
75%        66.000000        69.000000       67.000000      13.000000   
max        96.000000        95.000000       95.000000      94.000000   

         gk_handling     gk_kicking  gk_positioning    gk_reflexes  
count  183142.000000  183142.000000   183142.000000  183142.000000  
mean       16.063612      20.998362       16.132154      16.441439  
std        15.867382      21.452980       16.099175      17.198155  
min         1.000000       1.000000        1.000000       1.000000  
25%         8.000000       8.000000        8.000000       8.000000  
50%        11.000000      12.000000       11.000000      11.000000  
75%        15.000000      15.000000       15.000000      15.000000  
max        93.000000      97.000000       96.000000      96.000000  

[8 rows x 38 columns]

Among all other values, we see that the overall rating for players has a maximum of 94 and a minimum of 33 with a mean of 69. So many of the players are above average players. That is what we expect from the top leagues in Europe. The highest potential is 97 with a mean of 74 which also confirms our expectation. 

Now I will consider the player data frame. 


In [33]:
# checking dimensions
f'The dataframe has {player_df.shape[0]} samples and {player_df.shape[1]} columns'

'The dataframe has 11060 samples and 7 columns'

Information has been collected on 11,060 players in Europe for the period under consideration. This shows that there are two more players in the player attributes data frame (11,062) than we have in the player data frame. Maybe while compiling the data, the collectors did not insert information about two players in the player data frame. 

In [34]:
# checking for datatype consistency
player_df.dtypes

id                      int64
player_api_id           int64
player_name            object
player_fifa_api_id      int64
birthday               object
height                float64
weight                  int64
dtype: object

The only data inconsistency is in the birthday column. It was characterised as a string. We need to convert it to datetime object. 

In [35]:
# checking for nulls and missing data
player_df.isnull().sum()

id                    0
player_api_id         0
player_name           0
player_fifa_api_id    0
birthday              0
height                0
weight                0
dtype: int64

There are no missing data in the player dataframe. 

In [36]:
# check for duplicates
sum(player_df.duplicated())

0

There are no duplicates in the data frame. 

In [37]:
# check number of unique values in each column 
player_df.nunique()

id                    11060
player_api_id         11060
player_name           10848
player_fifa_api_id    11060
birthday               5762
height                   20
weight                   50
dtype: int64

We can see that some of the birthdays of players fall on the same day. Also, many players share similar heights and weights. That shows the players must be keeping to a strict and closely watched health regime. 

In [38]:
# descriptive statistics for the columns
player_df.describe()

id  player_api_id  player_fifa_api_id        height  \
count  11060.000000   11060.000000        11060.000000  11060.000000   
mean    5537.511392  156582.427215       165664.910488    181.867445   
std     3197.692647  160713.700624        58649.928360      6.369201   
min        1.000000    2625.000000            2.000000    157.480000   
25%     2767.750000   35555.500000       151889.500000    177.800000   
50%     5536.500000   96619.500000       184671.000000    182.880000   
75%     8306.250000  212470.500000       203883.250000    185.420000   
max    11075.000000  750584.000000       234141.000000    208.280000   

             weight  
count  11060.000000  
mean     168.380289  
std       14.990217  
min      117.000000  
25%      159.000000  
50%      168.000000  
75%      179.000000  
max      243.000000

The interquartile range for weight is 20. That shows most players follow a good health regime as to maintaining a good weight. 

Next is the team attributes data frame. 

In [39]:
# check the dimensions of the data 
f'There are {team_attributes_df.shape[0]} samples and {team_attributes_df.shape[1]} columns.'

'There are 1458 samples and 25 columns.'

In [40]:
# checking datatypes consistency
team_attributes_df.dtypes

id                                  int64
team_fifa_api_id                    int64
team_api_id                         int64
date                               object
buildUpPlaySpeed                    int64
buildUpPlaySpeedClass              object
buildUpPlayDribbling              float64
buildUpPlayDribblingClass          object
buildUpPlayPassing                  int64
buildUpPlayPassingClass            object
buildUpPlayPositioningClass        object
chanceCreationPassing               int64
chanceCreationPassingClass         object
chanceCreationCrossing              int64
chanceCreationCrossingClass        object
chanceCreationShooting              int64
chanceCreationShootingClass        object
chanceCreationPositioningClass     object
defencePressure                     int64
defencePressureClass               object
defenceAggression                   int64
defenceAggressionClass             object
defenceTeamWidth                    int64
defenceTeamWidthClass             

The only inconsistent data is the date column. We need to convert it to a datetime column. 

In [41]:
# check for nulls and missing data
team_attributes_df.isnull().sum()

id                                  0
team_fifa_api_id                    0
team_api_id                         0
date                                0
buildUpPlaySpeed                    0
buildUpPlaySpeedClass               0
buildUpPlayDribbling              969
buildUpPlayDribblingClass           0
buildUpPlayPassing                  0
buildUpPlayPassingClass             0
buildUpPlayPositioningClass         0
chanceCreationPassing               0
chanceCreationPassingClass          0
chanceCreationCrossing              0
chanceCreationCrossingClass         0
chanceCreationShooting              0
chanceCreationShootingClass         0
chanceCreationPositioningClass      0
defencePressure                     0
defencePressureClass                0
defenceAggression                   0
defenceAggressionClass              0
defenceTeamWidth                    0
defenceTeamWidthClass               0
defenceDefenderLineClass            0
dtype: int64

The only column with missing data is the buildUpPlayDribbling column. It will be left as is, and all the more so where there is no data dictionary to understand what the column means. 

In [42]:
# checking for duplicated rows
sum(team_attributes_df.duplicated())

0

There are no duplicate rows. 

In [43]:
# checking for unique values in each column
team_attributes_df.nunique()

id                                1458
team_fifa_api_id                   285
team_api_id                        288
date                                 6
buildUpPlaySpeed                    57
buildUpPlaySpeedClass                3
buildUpPlayDribbling                49
buildUpPlayDribblingClass            3
buildUpPlayPassing                  58
buildUpPlayPassingClass              3
buildUpPlayPositioningClass          2
chanceCreationPassing               50
chanceCreationPassingClass           3
chanceCreationCrossing              56
chanceCreationCrossingClass          3
chanceCreationShooting              57
chanceCreationShootingClass          3
chanceCreationPositioningClass       2
defencePressure                     48
defencePressureClass                 3
defenceAggression                   47
defenceAggressionClass               3
defenceTeamWidth                    43
defenceTeamWidthClass                3
defenceDefenderLineClass             2
dtype: int64

The difference in numbers between the team_fifa_api_id and team_api_id is intriguing. When investigating the team data frame below, we will get the actual figure for the number of teams in the data set. 

In [44]:
# descriptive statistics of the team attributes data frame
team_attributes_df.describe()

id  team_fifa_api_id    team_api_id  buildUpPlaySpeed  \
count  1458.000000       1458.000000    1458.000000       1458.000000   
mean    729.500000      17706.982167    9995.727023         52.462277   
std     421.032659      39179.857739   13264.869900         11.545869   
min       1.000000          1.000000    1601.000000         20.000000   
25%     365.250000        110.000000    8457.750000         45.000000   
50%     729.500000        485.000000    8674.000000         52.000000   
75%    1093.750000       1900.000000    9904.000000         62.000000   
max    1458.000000     112513.000000  274581.000000         80.000000   

       buildUpPlayDribbling  buildUpPlayPassing  chanceCreationPassing  \
count            489.000000         1458.000000            1458.000000   
mean              48.607362           48.490398              52.165295   
std                9.678290           10.896101              10.360793   
min               24.000000           20.000000              21.000000   
25%               42.000000           40.000000              46.000000   
50%               49.000000           50.000000              52.000000   
75%               55.000000           55.000000              59.000000   
max               77.000000           80.000000              80.000000   

       chanceCreationCrossing  chanceCreationShooting  defencePressure  \
count             1458.000000             1458.000000      1458.000000   
mean                53.731824               53.969136        46.017147   
std                 11.086796               10.327566        10.227225   
min                 20.000000               22.000000        23.000000   
25%                 47.000000               48.000000        39.000000   
50%                 53.000000               53.000000        45.000000   
75%                 62.000000               61.000000        51.000000   
max                 80.000000               80.000000        72.000000   

       defenceAggression  defenceTeamWidth  
count        1458.000000       1458.000000  
mean           49.251029         52.185871  
std             9.738028          9.574712  
min            24.000000         29.000000  
25%            44.000000         47.000000  
50%            48.000000         52.000000  
75%            55.000000         58.000000  
max            72.000000         73.000000

Next is the team data frame that will be investigated. 

In [45]:
# checking the dimensions
f'There are {team_df.shape[0]} samples and {team_df.shape[1]} columns'

'There are 299 samples and 5 columns'

So now we can confidently say there are 299 teams in the data set. 

In [46]:
# check datatypes consistency
team_df.dtypes

id                    int64
team_api_id           int64
team_fifa_api_id    float64
team_long_name       object
team_short_name      object
dtype: object

The data types are consistent. 

In [47]:
# check for nulls and missing data
team_df.isnull().sum()

id                   0
team_api_id          0
team_fifa_api_id    11
team_long_name       0
team_short_name      0
dtype: int64

Now we understand why the team_fifa_api_id column in the earlier dataframe, team_attributes, was different. There are some teams without the FIFA id. So, while using the dataset, we will not deal with the FIFA id but only use the team api id. 

In [48]:
# check for duplicate rows
sum(team_df.duplicated())

0

There are no duplicate rows. 

In [49]:
# check for the unique values in the data frame
team_df.nunique()

id                  299
team_api_id         299
team_fifa_api_id    285
team_long_name      296
team_short_name     259
dtype: int64

There is a possibility that in the dataframe, some team long names are repeated and some teams have similar short names. That is why these columns have lesser number of unique values than the team api id. 

### Data Cleaning

Based on the findings during the data wrangling stage, a thorough data cleaning will now be done. 

1. In player_attributes_df, we have to convert the date column from string to datetime object. 

In [53]:
# converting the date column from string to datetime object. 
player_attributes_df['date'] = pd.to_datetime(player_attributes_df['date'])

2. In player_df, the birthday column is characterized as a string. It also needs to be converted to datetime object. 

In [55]:
player_df['birthday'] = pd.to_datetime(player_df['birthday'])

3. In the team_attributes_df, the only inconsistent data is the date column. It has to be coinverted to datetime object. 

In [57]:
team_attributes_df['date'] = pd.to_datetime(team_attributes_df['date'])

## Exploratory Data Analysis 

We will explore the data using visual aids and statistics. To aid our exploration, we will concentrate our exploration on key questions we want to understand from the data. 



### Key Question 1: What teams won the most games?

We'll analyze this per country and then overall for the 9 years period. 

First, we'll need to merge the match_df with country_df tables so we can associate a country with a match. 

In [59]:
# match_df too big. Get relevant columns
match_cols = ["country_id", "league_id", "season", "stage", "date", 
"match_api_id", "home_team_api_id", "away_team_api_id", "home_team_goal", "away_team_goal"]
match_filtered = match_df[match_cols]

In [60]:
match_filtered.head()

country_id  league_id     season  stage                 date  match_api_id  \
0           1          1  2008/2009      1  2008-08-17 00:00:00        492473   
1           1          1  2008/2009      1  2008-08-16 00:00:00        492474   
2           1          1  2008/2009      1  2008-08-16 00:00:00        492475   
3           1          1  2008/2009      1  2008-08-17 00:00:00        492476   
4           1          1  2008/2009      1  2008-08-16 00:00:00        492477   

   home_team_api_id  away_team_api_id  home_team_goal  away_team_goal  
0              9987              9993               1               1  
1             10000              9994               0               0  
2              9984              8635               0               3  
3              9991              9998               5               0  
4              7947              9985               1               3

In [61]:
# the merge
match_country = match_filtered.merge(country_df, left_on="country_id", right_on="id")
match_country.head()

country_id  league_id     season  stage                 date  match_api_id  \
0           1          1  2008/2009      1  2008-08-17 00:00:00        492473   
1           1          1  2008/2009      1  2008-08-16 00:00:00        492474   
2           1          1  2008/2009      1  2008-08-16 00:00:00        492475   
3           1          1  2008/2009      1  2008-08-17 00:00:00        492476   
4           1          1  2008/2009      1  2008-08-16 00:00:00        492477   

   home_team_api_id  away_team_api_id  home_team_goal  away_team_goal  id  \
0              9987              9993               1               1   1   
1             10000              9994               0               0   1   
2              9984              8635               0               3   1   
3              9991              9998               5               0   1   
4              7947              9985               1               3   1   

      name  
0  Belgium  
1  Belgium  
2  Belgium  
3  Belgium  
4  Belgium

I will start filtering for English matches. 

In [63]:
english_matches = match_country[match_country['name'] == "England"]
english_matches.head()

country_id  league_id     season  stage                 date  \
1728        1729       1729  2008/2009      1  2008-08-17 00:00:00   
1729        1729       1729  2008/2009      1  2008-08-16 00:00:00   
1730        1729       1729  2008/2009      1  2008-08-16 00:00:00   
1731        1729       1729  2008/2009      1  2008-08-16 00:00:00   
1732        1729       1729  2008/2009      1  2008-08-17 00:00:00   

      match_api_id  home_team_api_id  away_team_api_id  home_team_goal  \
1728        489042             10260             10261               1   
1729        489043              9825              8659               1   
1730        489044              8472              8650               0   
1731        489045              8654              8528               2   
1732        489046             10252              8456               4   

      away_team_goal    id     name  
1728               1  1729  England  
1729               0  1729  England  
1730               1  1729  England  
1731               1  1729  England  
1732               2  1729  England

I will now pick out only the relevant columns for the analysis. Since I am taking a holistic view, the date column is not necessary. In another analysis, I might want to look at insights per year or season, but for now, I am considering the whole period. These helps see what teams have been in the top English Premiership for the 9 years since some teams can drop to the Championship. 

In [64]:
english_cols = ["home_team_api_id", "away_team_api_id", "home_team_goal", "away_team_goal", "name"]
english_matches_updated = english_matches[english_cols]
english_matches_updated

home_team_api_id  away_team_api_id  home_team_goal  away_team_goal  \
1728             10260             10261               1               1   
1729              9825              8659               1               0   
1730              8472              8650               0               1   
1731              8654              8528               2               1   
1732             10252              8456               4               2   
...                ...               ...             ...             ...   
4763              8466              8197               2               2   
4764             10003             10194               0               1   
4765              8586              8650               0               0   
4766              9817              9825               0               3   
4767              8659              8472               1               0   

         name  
1728  England  
1729  England  
1730  England  
1731  England  
1732  England  
...       ...  
4763  England  
4764  England  
4765  England  
4766  England  
4767  England  

[3040 rows x 5 columns]

I will now insert the names for the home team and away team. 

In [78]:
def insert_team_name(x):
    # get the x value in team_api_id of team_df and return the row
    team_row = team_df.loc[team_df['team_api_id'] == x]
    # get the long name for that row
    long_name = team_row['team_long_name'].values
    result = "".join(long_name)
    return result

In [79]:
english_matches_updated['home_team'] = english_matches_updated['home_team_api_id'].apply(insert_team_name)

/tmp/ipykernel_233/1709332459.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  english_matches_updated['home_team'] = english_matches_updated['home_team_api_id'].apply(insert_team_name)


In [80]:
english_matches_updated.head()

home_team_api_id  away_team_api_id  home_team_goal  away_team_goal  \
1728             10260             10261               1               1   
1729              9825              8659               1               0   
1730              8472              8650               0               1   
1731              8654              8528               2               1   
1732             10252              8456               4               2   

         name          home_team  
1728  England  Manchester United  
1729  England            Arsenal  
1730  England         Sunderland  
1731  England    West Ham United  
1732  England        Aston Villa

In [81]:
english_matches_updated['away_team'] = english_matches_updated['away_team_api_id'].apply(insert_team_name)

/tmp/ipykernel_233/2221444958.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  english_matches_updated['away_team'] = english_matches_updated['away_team_api_id'].apply(insert_team_name)


In [82]:
english_matches_updated.head()

home_team_api_id  away_team_api_id  home_team_goal  away_team_goal  \
1728             10260             10261               1               1   
1729              9825              8659               1               0   
1730              8472              8650               0               1   
1731              8654              8528               2               1   
1732             10252              8456               4               2   

         name          home_team             away_team  
1728  England  Manchester United      Newcastle United  
1729  England            Arsenal  West Bromwich Albion  
1730  England         Sunderland             Liverpool  
1731  England    West Ham United        Wigan Athletic  
1732  England        Aston Villa       Manchester City

In [83]:
english_updated_cols = ["name", "home_team", "away_team","home_team_goal", "away_team_goal"]
english_matches_filtered = english_matches_updated[english_updated_cols]

In [84]:
english_matches_filtered.head()

name          home_team             away_team  home_team_goal  \
1728  England  Manchester United      Newcastle United               1   
1729  England            Arsenal  West Bromwich Albion               1   
1730  England         Sunderland             Liverpool               0   
1731  England    West Ham United        Wigan Athletic               2   
1732  England        Aston Villa       Manchester City               4   

      away_team_goal  
1728               1  
1729               0  
1730               1  
1731               1  
1732               2

We'll now state the outcomes for the home and away teams. 

In [114]:
def home_team_outcome(x):
    if x['home_team_goal'] > x['away_team_goal']:
        return "Win"
    elif x['home_team_goal'] == x['away_team_goal']:
        return "Draw"
    else: 
        return "Loss"         


In [115]:
english_matches_filtered['home_team_outcome'] = english_matches_filtered.apply(home_team_outcome, axis=1)

/tmp/ipykernel_233/2336430521.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  english_matches_filtered['home_team_outcome'] = english_matches_filtered.apply(home_team_outcome, axis=1)


In [116]:
english_matches_filtered.head()

name          home_team             away_team  home_team_goal  \
1728  England  Manchester United      Newcastle United               1   
1729  England            Arsenal  West Bromwich Albion               1   
1730  England         Sunderland             Liverpool               0   
1731  England    West Ham United        Wigan Athletic               2   
1732  England        Aston Villa       Manchester City               4   

      away_team_goal home_team_outcome  
1728               1              Draw  
1729               0               Win  
1730               1              Loss  
1731               1               Win  
1732               2               Win

In [117]:
def away_team_outcome(x):
    if x['away_team_goal'] > x['home_team_goal']:
        return "Win"
    elif x['away_team_goal'] == x['home_team_goal']:
        return "Draw"
    else: 
        return "Loss"         


In [118]:
english_matches_filtered['away_team_outcome'] = english_matches_filtered.apply(away_team_outcome, axis=1)

In [119]:
english_matches_filtered.head()

name          home_team             away_team  home_team_goal  \
1728  England  Manchester United      Newcastle United               1   
1729  England            Arsenal  West Bromwich Albion               1   
1730  England         Sunderland             Liverpool               0   
1731  England    West Ham United        Wigan Athletic               2   
1732  England        Aston Villa       Manchester City               4   

      away_team_goal home_team_outcome away_team_outcome  
1728               1              Draw              Draw  
1729               0               Win              Loss  
1730               1              Loss               Win  
1731               1               Win              Loss  
1732               2               Win              Loss

## References

1. [How to extract tables from SQLite Database. Stackoverflow post.](https://stackoverflow.com/questions/18827028/write-to-csv-from-sqlite3-database-in-python)